File: python/pyfory/meta/typedef_encoder.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.


from pyfory.meta.typedef import (
    FieldInfo,
    TypeDef,
    build_field_infos,
    SMALL_NUM_FIELDS_THRESHOLD,
    REGISTER_BY_NAME_FLAG,
    COMPATIBLE_TYPEDEF_FLAG,
    STRUCT_TYPEDEF_FLAG,
    FIELD_NAME_SIZE_THRESHOLD,
    BIG_NAME_THRESHOLD,
    COMPRESS_META_FLAG,
    META_SIZE_MASKS,
    FIELD_NAME_ENCODINGS,
    NAMESPACE_ENCODINGS,
    TYPE_NAME_ENCODINGS,
    FIELD_NAME_ENCODING_TAG_ID,
    TAG_ID_SIZE_THRESHOLD,
    is_struct_typedef_kind,
    xlang_non_struct_kind_code,
    _typedef_header_hash,
)
from pyfory.meta.metastring import MetaStringEncoder
from pyfory._fory import NO_USER_TYPE_ID
from pyfory.types import TypeId

from pyfory.serialization import Buffer


# Meta string encoders
NAMESPACE_ENCODER = MetaStringEncoder(".", "_")
TYPENAME_ENCODER = MetaStringEncoder("$", "_")
FIELD_NAME_ENCODER = MetaStringEncoder("$", "_")


def encode_typedef(type_resolver, cls, include_fields: bool = True):
    """
    Encode the typedef of the type for xlang serialization.

    Args:
        type_resolver: The type resolver.
        cls: The class to encode.

    Returns:
        The encoded TypeDef.
    """
    type_id, user_type_id = type_resolver.get_registered_type_ids(cls)
    if include_fields and is_struct_typedef_kind(type_id):
        field_infos = build_field_infos(type_resolver, cls)
    else:
        field_infos = []

    buffer = Buffer.allocate(64)

    # Write kind header
    if is_struct_typedef_kind(type_id):
        num_fields = len(field_infos)
        header = STRUCT_TYPEDEF_FLAG | min(num_fields, SMALL_NUM_FIELDS_THRESHOLD)
        if type_id in {TypeId.COMPATIBLE_STRUCT, TypeId.NAMED_COMPATIBLE_STRUCT}:
            header |= COMPATIBLE_TYPEDEF_FLAG
        if type_resolver.is_registered_by_name(cls):
            header |= REGISTER_BY_NAME_FLAG
        if num_fields >= SMALL_NUM_FIELDS_THRESHOLD:
            buffer.write_uint8(header)
            buffer.write_var_uint32(num_fields - SMALL_NUM_FIELDS_THRESHOLD)
        else:
            buffer.write_uint8(header)
    else:
        if field_infos:
            raise ValueError(f"Non-struct TypeDef {type_id} cannot carry field metadata")
        buffer.write_uint8(xlang_non_struct_kind_code(type_id))

    # Write type info
    if type_resolver.is_registered_by_name(cls):
        namespace, typename = type_resolver.get_registered_name(cls)
        write_namespace(buffer, namespace)
        write_typename(buffer, typename)
    else:
        assert type_resolver.is_registered_by_id(cls=cls), "Class must be registered by name or id"
        if user_type_id in {None, NO_USER_TYPE_ID}:
            raise ValueError(f"user_type_id required for type_id {type_id}")
        buffer.write_var_uint32(user_type_id)

    # Write fields info
    write_fields_info(type_resolver, buffer, field_infos)

    # Get the encoded binary (only the written portion, not the full buffer)
    binary = buffer.to_bytes(0, buffer.get_writer_index())

    is_compressed = False
    # Prepend header
    binary = prepend_header(binary, is_compressed)
    # Extract namespace and typename
    if type_resolver.is_registered_by_name(cls):
        namespace, typename = type_resolver.get_registered_name(cls)
    else:
        splits = cls.__name__.rsplit(".", 1)
        if len(splits) == 1:
            splits.insert(0, "")
        namespace, typename = splits

    result = TypeDef(
        namespace,
        typename,
        cls,
        type_id,
        field_infos,
        binary,
        is_compressed,
        user_type_id=user_type_id,
    )
    return result


def prepend_header(buffer: bytes, is_compressed: bool):
    """Prepend header to the buffer."""
    meta_size = len(buffer)
    header_low_bits = min(meta_size, META_SIZE_MASKS)
    if is_compressed:
        header_low_bits |= COMPRESS_META_FLAG

    header = _typedef_header_hash(buffer, header_low_bits) | header_low_bits
    if header >= (1 << 63):
        header -= 1 << 64
    result = Buffer.allocate(meta_size + 8)
    result.write_int64(header)
    if meta_size >= META_SIZE_MASKS:
        result.write_var_uint32(meta_size - META_SIZE_MASKS)

    result.write_bytes(buffer)
    return result.to_bytes(0, result.get_writer_index())


def write_namespace(buffer: Buffer, namespace: str):
    """Write namespace using meta string encoding."""
    # - Package name encoding(omitted when class is registered):
    #    - encoding algorithm: `UTF_8/ALL_TO_LOWER_SPECIAL/LOWER_UPPER_DIGIT_SPECIAL`
    #    - Header: `6 bits size | 2 bits encoding flags`.
    #      The `6 bits size: 0~63`  will be used to indicate size `0~62`,
    #      the value `63` the size need more byte to read, the encoding will encode `size - 62` as a varint next.
    meta_string = NAMESPACE_ENCODER.encode(namespace, NAMESPACE_ENCODINGS)
    write_meta_string(buffer, meta_string, NAMESPACE_ENCODINGS.index(meta_string.encoding))


def write_typename(buffer: Buffer, typename: str):
    """Write typename using meta string encoding."""
    # - Class name encoding(omitted when class is registered):
    #     - encoding algorithm:
    # `UTF_8/LOWER_UPPER_DIGIT_SPECIAL/FIRST_TO_LOWER_SPECIAL/ALL_TO_LOWER_SPECIAL`
    #     - header: `6 bits size | 2 bits encoding flags`.
    #       The `6 bits size: 0~63`  will be used to indicate size `1~64`,
    #       the value `63` the size need more byte to read, the encoding will encode `size - 63` as a varint next.
    meta_string = TYPENAME_ENCODER.encode(typename, TYPE_NAME_ENCODINGS)
    write_meta_string(buffer, meta_string, TYPE_NAME_ENCODINGS.index(meta_string.encoding))


def write_meta_string(buffer: Buffer, meta_string, encoding_value: int):
    """Write a big meta string (namespace/typename) to the buffer using 6-bit size field."""
    # Write encoding and length combined in first byte
    length = len(meta_string.encoded_data)

    if length >= BIG_NAME_THRESHOLD:
        # Use threshold value and write additional length
        header = (BIG_NAME_THRESHOLD << 2) | encoding_value
        buffer.write_uint8(header)
        buffer.write_var_uint32(length - BIG_NAME_THRESHOLD)
    else:
        # Combine length and encoding in single byte
        header = (length << 2) | encoding_value
        buffer.write_uint8(header)

    # Write encoded data
    if meta_string.encoded_data:
        buffer.write_bytes(meta_string.encoded_data)


def write_fields_info(type_resolver, buffer: Buffer, field_infos: list):
    """Write field information to the buffer."""
    for field_info in field_infos:
        write_field_info(buffer, field_info)


def write_field_info(buffer: Buffer, field_info: FieldInfo):
    """Write a single field info to the buffer.

    Field header format (8 bits) - aligned with Java TypeDefEncoder (for xlang):
    - bit 0: ref tracking flag
    - bit 1: nullable flag
    - bits 2-5: size (4 bits, 0-14 inline, 15 = overflow)
    - bits 6-7: encoding type (0b00-10 = field name, 0b11 = TAG_ID)

    For TAG_ID encoding (when encoding = 0b11):
    - size field contains tag_id (0-14 inline, 15 = overflow)
    - No field name bytes to write

    For field name encoding (when encoding = 0b00-10):
    - size field contains (encoded_size - 1)
    - type info followed by field name meta string
    """
    # Build header flags (bits 0-1)
    header = 0
    if field_info.field_type.is_tracking_ref:
        header |= 0b01  # bit 0
    if field_info.field_type.is_nullable:
        header |= 0b10  # bit 1

    if field_info.uses_tag_id():
        # TAG_ID encoding (encoding = 0b11 at bits 6-7)
        tag_id = field_info.tag_id
        header |= FIELD_NAME_ENCODING_TAG_ID << 6  # encoding at bits 6-7

        if tag_id >= TAG_ID_SIZE_THRESHOLD:
            # Overflow: use 0b1111 and write extra varint
            header |= TAG_ID_SIZE_THRESHOLD << 2
            buffer.write_uint8(header)
            buffer.write_var_uint32(tag_id - TAG_ID_SIZE_THRESHOLD)
        else:
            # Inline tag_id (0-14)
            header |= tag_id << 2
            buffer.write_uint8(header)

        # Write field type info (no field name for TAG_ID)
        field_info.field_type.write(buffer, False)
    else:
        # Field name encoding
        encoding = FIELD_NAME_ENCODER.compute_encoding(field_info.name, FIELD_NAME_ENCODINGS)
        meta_string = FIELD_NAME_ENCODER.encode_with_encoding(field_info.name, encoding)
        # Store (length - 1) in size field, matching Java TypeDefEncoder
        field_name_binary_size = len(meta_string.encoded_data) - 1
        encoding_flags = FIELD_NAME_ENCODINGS.index(meta_string.encoding)
        header |= encoding_flags << 6  # encoding at bits 6-7

        if field_name_binary_size >= FIELD_NAME_SIZE_THRESHOLD:
            header |= FIELD_NAME_SIZE_THRESHOLD << 2
            buffer.write_uint8(header)
            buffer.write_var_uint32(field_name_binary_size - FIELD_NAME_SIZE_THRESHOLD)
        else:
            header |= field_name_binary_size << 2
            buffer.write_uint8(header)

        # Write field type info BEFORE field name (matching Java TypeDefEncoder order)
        field_info.field_type.write(buffer, False)

        # Write field name meta string
        buffer.write_bytes(meta_string.encoded_data)

File: java/fory-core/src/main/java/org/apache/fory/meta/TypeDefEncoder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.meta;

import static org.apache.fory.meta.Encoders.fieldNameEncodingsList;
import static org.apache.fory.meta.NativeTypeDefEncoder.prependHeader;
import static org.apache.fory.meta.NativeTypeDefEncoder.writePkgName;
import static org.apache.fory.meta.NativeTypeDefEncoder.writeTypeName;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import org.apache.fory.logging.Logger;
import org.apache.fory.logging.LoggerFactory;
import org.apache.fory.memory.MemoryBuffer;
import org.apache.fory.meta.FieldTypes.FieldType;
import org.apache.fory.resolver.TypeInfo;
import org.apache.fory.resolver.TypeResolver;
import org.apache.fory.resolver.XtypeResolver;
import org.apache.fory.type.Descriptor;
import org.apache.fory.type.DescriptorGrouper;
import org.apache.fory.type.TypeUtils;
import org.apache.fory.type.Types;
import org.apache.fory.util.Preconditions;
import org.apache.fory.util.StringUtils;
import org.apache.fory.util.Utils;

/**
 * An encoder which encode {@link TypeDef} into binary. See spec documentation:
 * docs/specification/fory_xlang_serialization_spec.md <a
 * href="https://fory.apache.org/docs/specification/fory_xlang_serialization_spec">...</a>
 */
class TypeDefEncoder {
  private static final Logger LOG = LoggerFactory.getLogger(TypeDefEncoder.class);
  private static final java.util.function.Function<Descriptor, Descriptor> IDENTITY_DESCRIPTOR =
      descriptor -> descriptor;

  /** Build class definition from fields of class. */
  static TypeDef buildTypeDef(XtypeResolver resolver, Class<?> type) {
    List<Descriptor> fieldDescriptors =
        dropShadowedXlangNameFields(type, resolver.getFieldDescriptors(type, true));
    DescriptorGrouper descriptorGrouper =
        resolver.groupDescriptors(fieldDescriptors, false, IDENTITY_DESCRIPTOR);
    TypeInfo typeInfo = resolver.getTypeInfo(type);
    List<Descriptor> descriptors;
    int typeId = typeInfo.getTypeId();
    if (Types.isStructType(typeId)) {
      descriptors = descriptorGrouper.getSortedDescriptors();
    } else {
      descriptors = new ArrayList<>();
    }
    return buildTypeDefWithFieldInfos(
        resolver, type, buildFieldsInfoFromDescriptors(resolver, type, descriptors));
  }

  private static List<Descriptor> dropShadowedXlangNameFields(
      Class<?> type, List<Descriptor> descriptors) {
    List<Descriptor> result = new ArrayList<>(descriptors.size());
    Map<String, Integer> nameFieldIndex = new HashMap<>();
    for (Descriptor descriptor : descriptors) {
      if (descriptor.hasForyFieldId()) {
        result.add(descriptor);
        continue;
      }
      String fieldIdentifier = descriptor.getSnakeCaseName();
      Integer previousIndex = nameFieldIndex.get(fieldIdentifier);
      if (previousIndex == null) {
        nameFieldIndex.put(fieldIdentifier, result.size());
        result.add(descriptor);
        continue;
      }
      Descriptor previous = result.get(previousIndex);
      int distance = inheritanceDistance(type, descriptor);
      int previousDistance = inheritanceDistance(type, previous);
      if (!descriptor.getName().equals(previous.getName()) || distance == previousDistance) {
        throw new IllegalArgumentException(
            "Duplicate xlang field identifier "
                + fieldIdentifier
                + " for fields "
                + previous.getName()
                + " and "
                + descriptor.getName()
                + " in class "
                + type.getName()
                + ". Configure explicit non-negative field IDs to disambiguate them.");
      }
      // Untagged xlang fields are addressed by snake_case identifier. Only a true Java field hide
      // across the inheritance chain can be represented by dropping the farther inherited field.
      if (distance < previousDistance) {
        result.set(previousIndex, descriptor);
      }
    }
    return result;
  }

  private static int inheritanceDistance(Class<?> type, Descriptor descriptor) {
    String declaringClass = descriptor.getDeclaringClass();
    int distance = 0;
    for (Class<?> current = type; current != null; current = current.getSuperclass()) {
      if (current.getName().equals(declaringClass)) {
        return distance;
      }
      distance++;
    }
    return Integer.MAX_VALUE;
  }

  static List<FieldInfo> buildFieldsInfo(TypeResolver resolver, Class<?> type, List<Field> fields) {
    List<Descriptor> descriptors = new ArrayList<>(fields.size());
    for (Field field : fields) {
      descriptors.add(new Descriptor(field, TypeUtils.getFieldTypeRef(field), null, null));
    }
    return buildFieldsInfoFromDescriptors(resolver, type, descriptors);
  }

  static List<FieldInfo> buildFieldsInfoFromDescriptors(
      TypeResolver resolver, Class<?> type, List<Descriptor> descriptors) {
    Set<Integer> usedTagIds = new HashSet<>();
    return descriptors.stream()
        .map(
            descriptor -> {
              FieldType fieldType = FieldTypes.buildFieldType(resolver, descriptor);
              if (descriptor.hasForyField()) {
                int tagId = descriptor.getForyFieldId();
                if (tagId >= 0) {
                  if (!usedTagIds.add(tagId)) {
                    throw new IllegalArgumentException(
                        "Duplicate tag id "
                            + tagId
                            + " for field "
                            + descriptor.getName()
                            + " in class "
                            + type.getName());
                  }
                  return new FieldInfo(
                      type.getName(), descriptor.getName(), fieldType, (short) tagId);
                }
                // Negative is the annotation default sentinel for no configured tag ID; fall
                // through to create regular FieldInfo. User-facing tag IDs must be non-negative.
              }
              return new FieldInfo(type.getName(), descriptor.getName(), fieldType);
            })
        .collect(Collectors.toList());
  }

  static TypeDef buildTypeDefWithFieldInfos(
      XtypeResolver resolver, Class<?> type, List<FieldInfo> fieldInfos) {
    fieldInfos = new ArrayList<>(fieldInfos);
    TypeInfo typeInfo = resolver.getTypeInfo(type);
    MemoryBuffer encodeTypeDef = encodeTypeDef(resolver, type, fieldInfos);
    byte[] typeDefBytes = encodeTypeDef.getBytes(0, encodeTypeDef.writerIndex());
    TypeDef typeDef =
        new TypeDef(
            new ClassSpec(type, typeInfo.getTypeId(), typeInfo.getUserTypeId()),
            fieldInfos,
            encodeTypeDef.getInt64(0),
            typeDefBytes);
    if (Utils.DEBUG_OUTPUT_ENABLED) {
      LOG.info("[Java TypeDef BUILT] " + typeDef);
    }
    return typeDef;
  }

  static final int SMALL_NUM_FIELDS_THRESHOLD = 0b11111;
  static final int REGISTER_BY_NAME_FLAG = 0b0010_0000;
  static final int COMPATIBLE_FLAG = 0b0100_0000;
  static final int STRUCT_FLAG = 0b1000_0000;
  static final int FIELD_NAME_SIZE_THRESHOLD = 0b1111;

  // see spec documentation: docs/specification/xlang_serialization_spec.md
  // https://fory.apache.org/docs/specification/fory_xlang_serialization_spec
  static MemoryBuffer encodeTypeDef(XtypeResolver resolver, Class<?> type, List<FieldInfo> fields) {
    TypeInfo typeInfo = resolver.getTypeInfo(type);
    int typeId = typeInfo.getTypeId();
    boolean isStruct = Types.isStructType(typeId);
    Preconditions.checkArgument(
        isStruct || fields.isEmpty(), "Non-struct TypeDef %s cannot carry field metadata", typeId);
    MemoryBuffer buffer = MemoryBuffer.newHeapBuffer(128);
    buffer.writeByte(-1); // placeholder for header, update later
    if (isStruct) {
      int fieldCount = fields.size();
      int currentClassHeader = STRUCT_FLAG | Math.min(fieldCount, SMALL_NUM_FIELDS_THRESHOLD);
      if (typeId == Types.COMPATIBLE_STRUCT || typeId == Types.NAMED_COMPATIBLE_STRUCT) {
        currentClassHeader |= COMPATIBLE_FLAG;
      }
      if (fieldCount >= SMALL_NUM_FIELDS_THRESHOLD) {
        buffer.writeVarUInt32(fieldCount - SMALL_NUM_FIELDS_THRESHOLD);
      }
      if (resolver.isRegisteredById(type)) {
        Preconditions.checkArgument(
            typeInfo.getUserTypeId() != -1,
            "User type id is required for typeId %s",
            typeInfo.getTypeId());
        buffer.writeVarUInt32(typeInfo.getUserTypeId());
      } else {
        Preconditions.checkArgument(resolver.isRegisteredByName(type));
        currentClassHeader |= REGISTER_BY_NAME_FLAG;
        String ns = typeInfo.decodeNamespace();
        String typename = typeInfo.decodeTypeName();
        writePkgName(buffer, ns);
        writeTypeName(buffer, typename);
      }
      buffer.putByte(0, currentClassHeader);
      writeFieldsInfo(resolver, buffer, fields);
    } else {
      buffer.putByte(0, nonStructKindCode(typeId));
      if (resolver.isRegisteredById(type)) {
        Preconditions.checkArgument(
            typeInfo.getUserTypeId() != -1,
            "User type id is required for typeId %s",
            typeInfo.getTypeId());
        buffer.writeVarUInt32(typeInfo.getUserTypeId());
      } else {
        Preconditions.checkArgument(resolver.isRegisteredByName(type));
        String ns = typeInfo.decodeNamespace();
        String typename = typeInfo.decodeTypeName();
        writePkgName(buffer, ns);
        writeTypeName(buffer, typename);
      }
    }
    return prependHeader(buffer, false);
  }

  static int nonStructKindCode(int typeId) {
    switch (typeId) {
      case Types.ENUM:
        return 0;
      case Types.NAMED_ENUM:
        return 1;
      case Types.EXT:
        return 2;
      case Types.NAMED_EXT:
        return 3;
      case Types.TYPED_UNION:
        return 4;
      case Types.NAMED_UNION:
        return 5;
      default:
        throw new IllegalArgumentException("Unsupported TypeDef kind " + typeId);
    }
  }

  /** Write field type and name info. Every field info format: `header + type info + field name` */
  static void writeFieldsInfo(XtypeResolver resolver, MemoryBuffer buffer, List<FieldInfo> fields) {
    for (FieldInfo fieldInfo : fields) {
      FieldType fieldType = fieldInfo.getFieldType();
      // header: 2 bits field name encoding + 4 bits size + nullability flag + ref tracking flag
      int header = ((fieldType.trackingRef() ? 1 : 0));
      header |= fieldType.nullable() ? 0b10 : 0b00;
      int size, encodingFlags;
      byte[] encoded = null;
      if (fieldInfo.hasFieldId()) {
        size = fieldInfo.getFieldId();
        encodingFlags = 3;
      } else {
        // Convert camelCase field names to snake_case for xlang interoperability
        String fieldName = StringUtils.lowerCamelToLowerUnderscore(fieldInfo.getFieldName());
        MetaString metaString = Encoders.encodeFieldName(fieldName);
        // Encoding `UTF_8/ALL_TO_LOWER_SPECIAL/LOWER_UPPER_DIGIT_SPECIAL/TAG_ID`
        encodingFlags = fieldNameEncodingsList.indexOf(metaString.getEncoding());
        encoded = metaString.getBytes();
        size = (encoded.length - 1);
      }
      header |= (byte) (encodingFlags << 6);
      boolean bigSize = size >= FIELD_NAME_SIZE_THRESHOLD;
      if (bigSize) {
        header |= 0b00111100;
        buffer.writeByte(header);
        buffer.writeVarUInt32Small7(size - FIELD_NAME_SIZE_THRESHOLD);
      } else {
        header |= (size << 2);
        buffer.writeByte(header);
      }
      fieldType.writeCrossLanguage(buffer, false);
      // write field name
      if (!fieldInfo.hasFieldId()) {
        buffer.writeBytes(encoded);
      }
    }
  }
}
